utils/ipfs.py

```diff
@@ -0,0 +1,222 @@
+
+
+# fetching url status
+async def _fetch_status(session, url: str) -> int:
+    async with session.get(url) as response:
+        return response.status
+
+
+# fetch cid from different api concurrency
+async def _fetch_cid_status_from_ipfs_async(cid: str) -> int:
+    if not cid:
+        return 404
+
+    request_urls = [
+        f"https://{cid}.ipfs.w3s.link",  # faster for uploaded files
+        f"https://api.web3.storage/status/{cid}",  # much faster for not uploaded files
+    ]
+
+    async with aiohttp.ClientSession() as session:
+        tasks = [asyncio.create_task(_fetch_status(session, url)) for url in request_urls]
+        done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+
+        for task in pending:
+            task.cancel()
+
+        for task in done:
+            return task.result()
+
+
+def verify_ipfs_description(text: str) -> list[Tuple[str, str]]:
+    messages: list[Tuple[str, str]] = []
+    if not text:
+        messages.append(
+            (
+                "error",
+                (
+                    "You provided an empty string as description. "
+                    "If you provide text as a description, it will help users to make a voting decision. "
+                    "The description is stored on the IPFS network and does not take up much space in the script."
+                ),
+            )
+        )
+
+    address_raw_groups = re.findall(rf"([^`]{REG_ETH_ADDRESS}|{REG_ETH_ADDRESS}[^`])", f" {text} ")
+    if address_raw_groups:
+        address_raw = list(map(lambda x: x[1] or x[2], address_raw_groups))
+        messages.append(
+            (
+                "warning",
+                (
+                    "You have wallet addresses in description which has no Markdown style. "
+                    "You could use inline code block to make it looks better. "
+                    "You need to add '`' before and after the address. Here is the list of addresses:\n"
+                    f"{linesep.join(address_raw)}"
+                ),
+            )
+        )
+
+    cid_raw_groups = re.findall(rf"([^`]{REG_CID_DEFAULT}|{REG_CID_DEFAULT}[^`])", f" {text} ")
+    if cid_raw_groups:
+        cid_raw = list(map(lambda x: x[1] or x[2], cid_raw_groups))
+        messages.append(
+            (
+                "warning",
+                (
+                    "You have CIDs in description which has no Markdown style. "
+                    "You could use inline code block to make it looks better. "
+                    "You need to add '`' before and after CID. Here is the list of CID:\n"
+                    f"{linesep.join(cid_raw)}"
+                ),
+            )
+        )
+    return messages
+
+
+def fetch_cid_status_from_ipfs(cid: str):
+    return asyncio.run(_fetch_cid_status_from_ipfs_async(cid))
+
+
+# calc_only could be using for test runs to avoid uploading
+def upload_vote_description_to_ipfs(text: str, calc_only: bool = False) -> IPFSUploadResult:
```

Comment:
Can we split this public func for clients into two counterparts: one for upload and one for calc only? would be more semantically correct, I believe